src/trusted/service_runtime/nacl_app_thread.c

```diff
@@ -79,46 +72,40 @@ struct NaClApp *NaClChildNapCtor(struct NaClApp *nap) {
   nap_child->user_entry_pt = nap_parent->user_entry_pt;
   nap_child->parent_id = nap_parent->cage_id;
   nap_child->parent = nap_parent;
-  nap_child->master = nap_master;
   nap_child->in_fork = 0;
 
-  /* avoid incrementing child count twice */
-  if (nap_master == nap_parent) {
-    nap_arr[0] = NULL;
+  for(char const *const *ce = nap_parent->clean_environ; ce && *ce; ++ce) {
+    envc++;
   }
-
-  /* don't lock master twice */
-  NaClXMutexLock(&nap_master->children_mu);
-  if (nap_parent != nap_master) {
-    NaClXMutexLock(&nap_parent->children_mu);
+  childe = malloc((envc + 1) * sizeof(char*));
+  nap_child->clean_environ = (const char *const *)childe;
+  for(char const *const *ce = nap_parent->clean_environ; ce && *ce; ++ce) {
+     *childe++ = *ce;
   }
+  *childe++ = NULL;
+
+  NaClXMutexLock(&ccmut);
+  cagecount++;
+  NaClXMutexUnlock(&ccmut);
+
+  NaClXMutexLock(&nap_parent->children_mu);
   /*
    * increment fork generation count and generate child
    * cage_id (both master and parent mutexes need to be held
    */
-  InitializeCage(nap_child, nap_master->cage_id + ++fork_num);
+  InitializeCage(nap_child, 1 +  ++fork_num);
```

Comment:
Should we be fixing this to start from cage 0 at this point? I think the "mystery" Cage 0 was fixed.

Reply:
I mean in posix, the init process has a pid 1, so I think 1 being the first pid is fine.

Reply:
Hmm, maybe we should make 1 a define here for clarity then.